Replace `QRegExp` with `QRegularExpression`

w0xlt opened the pull request:
Picking up https://github.com/bitcoin-core/gui/pull/606 (labeled "Up for grabs") and applying https://github.com/bitcoin-core/gui/pull/606#pullrequestreview-984607067 and https://github.com/bitcoin-core/gui/pull/606#issuecomment-1137149907.

Replaces occurrences of `QRegExp` usage with `QRegularExpression` as part of the roadmap for Qt6 integration.

Fixes https://github.com/bitcoin-core/gui/issues/578

hebasto commented:
@w0xlt 

Thank you for picking it up!

What is your opinion about https://github.com/bitcoin-core/gui/pull/606#issuecomment-1138795241?

w0xlt commented:
> What is your opinion about https://github.com/bitcoin-core/gui/pull/606#issuecomment-1138795241?

That looks good to me. I'll take a look at it.
Would you suggest a class/file to allocate the utility function?

hebasto commented:
> > What is your opinion about [#606 (comment)](https://github.com/bitcoin-core/gui/pull/606#issuecomment-1138795241)?
> 
> That looks good to me. I'll take a look at it. Would you suggest a class/file to allocate the utility function?

`namespace GUIUtil` in `qt/guiutil.{h,cpp}`?

hebasto commented on src/qt/test/optiontests.cpp:
```diff
@@ -122,3 +123,19 @@ void OptionTests::parametersInteraction()
+
+void OptionTests::extractFilter()
+{
+    QString filter = QString("Partially Signed Transaction (Binary) (*.psbt)");
+    QCOMPARE(GUIUtil::extractFirstSuffixFromFilter(filter), "psbt");
+
+    filter = QString("Comma separated file (*.csv)");
+    QCOMPARE(GUIUtil::extractFirstSuffixFromFilter(filter), "csv");
+
+    filter = QString("PNG Image (*.png)");
+    QCOMPARE(GUIUtil::extractFirstSuffixFromFilter(filter), "png");
+
+    filter = QString("Wallet Data (*.dat)");
+    QCOMPARE(GUIUtil::extractFirstSuffixFromFilter(filter), "dat");
```
Why 4 tests with the only suffix? Maybe just two tests? First test with one suffix, another one with two suffixes.

hebasto commented on src/qt/test/optiontests.cpp:
```diff
@@ -122,3 +123,19 @@ void OptionTests::parametersInteraction()
+
+    filter = QString("Wallet Data (*.dat)");
+    QCOMPARE(GUIUtil::extractFirstSuffixFromFilter(filter), "dat");
+
```
Drop empty line?

w0xlt replied:
Done in https://github.com/bitcoin-core/gui/commit/67364ebe4c499eb8effe8dac11a5e3648f30c6c7.

hebasto commented on src/qt/utilitydialog.cpp:
```diff
@@ -22,7 +22,7 @@
 #include <QLabel>
 #include <QMainWindow>
-#include <QRegExp>
+#include <QRegularExpression>
```
Add `#include <QString>` as well (due to using `QStringLiteral`)?

w0xlt replied:
Done in https://github.com/bitcoin-core/gui/pull/620/commits/ace9af5688662a56d855bd26f36bc5c145b909e3.

hebasto reviewed (commented):
Approach ACK c0c8b33748637372ee5b267f6bbcce33a5380f4d.

Suggesting to split the first commit into separated ones:
- factoring out the `extractFirstSuffixFromFilter9()` function
- replacement `QRegExp` with `QRegularExpression`

Also please apply [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy) for each your commit.

hebasto commented on src/qt/guiutil.h:
```diff
@@ -123,6 +123,13 @@ namespace GUIUtil
     QString getDefaultDataDirectory();
 
+
+    /** Extract first suffix from filter pattern "Description (*.foo)" or "Description (*.foo *.bar ...).
+
+      @param[in] filter Filter specification such as "Comma Separated Files (*.csv)"
+    */
```
Please consult with [Coding Style (Doxygen-compatible comments)](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments).

w0xlt replied:
Done in https://github.com/bitcoin-core/gui/pull/620/commits/c378535e28e0c54353fde9c4a66cb3134a3ee742.

w0xlt commented:
@hebasto thanks for the review.
Force-pushed https://github.com/bitcoin-core/gui/pull/620/commits/67364ebe4c499eb8effe8dac11a5e3648f30c6c7 addressing your suggestions.

hebasto reviewed (approved):
ACK 67364ebe4c499eb8effe8dac11a5e3648f30c6c7

laanwj commented:
Code review and lightly tested ACK 7e4ae5ff38e149f469369887858e59c79bc332ae
